### DISCUSSION TITLE
Execution of embedded linux image failed: Load Error

### DISCUSSION THREAD

**Discussion by taylorjonl** [zbm-dev/zfsbootmenu#611] [Questions] [answered]
I have a Dell R730XD that I had setup with Ubuntu using ZFS boot menu.  When I had it setup it had the Perc H730P but from reading that card even in HBA mode isn't truly an HBA so I decided to get a Perc HBA330.  After the change it will no longer boot, I get the following errors:

Execution of embedded linux image failed: Load Error

I have tried reinstalling multiple times, these are the instructions I am using:

https://gist.github.com/taylorjonl/cf2961c0386bec8d83b3ae5c22bf9a6c

I was able to run the Ubuntu installer and setup with a software RAID1 array, so I know the HBA card can be used to boot from.  I also tried using refind instead of the direct EFI partition, but it has the same error.

Hoping I can get this working, otherwise my next choice is to use OpenSuse with Snapper.

**Comment by ahesford** [comment]
When using rEFInd, are you just trying to boot the ZBM EFI executable, or loading our separate kernel and initramfs components? rEFInd won't work around the problem some hardware has with the gummiboot stub used in the ZFSBootMenu image.

**Reply by taylorjonl** [reply]
Basically I replaced these commands:

```
apt install efibootmgr
efibootmgr -c -d "$DISK1" -p "1" -L "ZFSBootMenu (Backup)" -l '\EFI\ZBM\VMLINUZ-BACKUP.EFI'
efibootmgr -c -d "$DISK1" -p "1" -L "ZFSBootMenu" -l '\EFI\ZBM\VMLINUZ.EFI'
```

With:

```
apt install refind
refind-install
rm /boot/refind_linux.conf

cat << EOF > /boot/efi/EFI/ZBM/refind_linux.conf
"Boot default"  "quiet loglevel=0 zbm.skip"
"Boot to menu"  "quiet loglevel=0 zbm.show"
EOF
```

Do you have any docs that can explain what you mean?  I am pretty new to this and don't understand a lot of the terms you are using.  I actually prefer not to use rEFInd but attempted it as a troubleshooting step.

**Reply by zdykstra** [reply, marked as answer]
Downloads the [components](https://get.zfsbootmenu.org/tar.gz) release of ZFSBootMenu and extract the files into /boot/EFI/EFI/ZBM. rEFInd will then just load a kernel and initramfs for you, instead of the problematic EFI bundle. The boot process is then:

rEFInd > ZBM kernel/initramfs > BE kernel/initramfs.  

You can tweak timeouts for both so that it really won't impact startup speed very much.

**Reply by taylorjonl** [reply]
That worked, thank you.  I even got it to work without rEFInd using roughly this command:
```
efibootmgr -c -d "$DISK1" -p "1" -L "ZFSBootMenu" -l '\EFI\ZBM\vmlinuz-bootmenu' -u "ro initrd=\EFI\zbm\initramfs-bootmenu.img quiet"
```

**Reply by woutergrimme** [reply]
My generate-zbm would not generate an .EFI file (could not find EFI-stub, never solved that), but it was happy to generate the components (component -> Enabled: true). But for the life of me I could not figure out how to make that work with efibootmgr. This fixed it, thank you!!

`-u "ro initrd=\EFI\zbm\initramfs-bootmenu.img quiet"`

**Reply by slymattz** [reply]
Just for the record, I had the same issue with a MacMini 2012 (model no. A1347). It wouldn't load a bundled vmlinuz.EFI file, so I generated components, created an appropriate boot entry with efibootmgr and it worked like a charm.

**Comment by zdykstra** [comment]
You'll just need to extract the components build into a filesystem that rEFInd can read - your ESP is fine for that. That should work fine for you, as long as the kernel is properly paired with an initramfs.
